### DISCUSSION TITLE
Code optimizations

### DISCUSSION THREAD

**Discussion by omgitsaheadcrab** [encode/httpx#1924] [General] [answered]
How are code optimizations received usually?

For example, [is_success](https://github.com/encode/httpx/blob/da8f959af02e8d8dec69311ac8981fed90944573/httpx/_status_codes.py#L49) can be made ~2x as fast if you avoid making two comparisons by using integer division. For simple functions (with decent names/docstrings) I believe the loss of readability that usually comes with an optimization is acceptable. Thoughts?

**Comment by lovelydinosaur** [comment, marked as answer]
I think the only sensible approach is case-by-case.

Generally I'd optimise for readability over micro-optimisation, but for the *specific* case perhaps? they're two equally decent formulations...

```python
    @classmethod
    def is_success(cls, value: int) -> bool:
        """
        Returns `True` for 2xx status codes, `False` otherwise.
        """
        return 200 <= value <= 299
```

vs.

```python
    @classmethod
    def is_success(cls, value: int) -> bool:
        """
        Returns `True` for 2xx status codes, `False` otherwise.
        """
        return value // 100 == 2
```

Writing this now, and looking at them side by side I *guess* I'd tend to suggest just sticking with the first. But there's no specific rule we can apply when looking at that trade-off.
